src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -566,32 +566,99 @@ private ValueTask<FlushResult> WriteDataAndTrailersAsync(Http2Stream stream, in
+            if (endOfHeaders && dataToFrame.Length == currentSize)
+            {
+                _outgoingFrame.HeadersFlags |= Http2HeadersFrameFlags.END_HEADERS;
+            }
 
-        while (!done)
-        {
-            _outgoingFrame.PrepareContinuation(Http2ContinuationFrameFlags.NONE, streamId);
+            WriteHeaderUnsynchronized();
+            _outputWriter.Write(dataToFrame[..currentSize]);
+            dataToFrame = dataToFrame.Slice(currentSize);
+        }
+    }
 
-            done = HPackHeaderWriter.ContinueEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, out payloadLength);
+    private void FinishWritingHeadersUnsynchronized(int streamId, int payloadLength, HPackHeaderWriter.HeaderWriteResult writeResult)
+    {
+        Debug.Assert(payloadLength <= _maxFrameSize, "The initial payload lengths is written to _headerEncodingBuffer with size of _maxFrameSize");
+        var bufferSize = _maxFrameSize;
+        byte[]? largeHeaderBuffer = null;
+        Span<byte> buffer;
+        if (writeResult == HPackHeaderWriter.HeaderWriteResult.Done)
+        {
+            // Fast path, only a single HEADER frame.
             _outgoingFrame.PayloadLength = payloadLength;
-
-            if (done)
+            _outgoingFrame.HeadersFlags |= Http2HeadersFrameFlags.END_HEADERS;
+            WriteHeaderUnsynchronized();
+            _outputWriter.Write(_headerEncodingBuffer.AsSpan(0, payloadLength));
+            return;
+        }
+        else if (writeResult == HPackHeaderWriter.HeaderWriteResult.MoreHeaders)
+        {
+            _outgoingFrame.PayloadLength = payloadLength;
+            WriteHeaderUnsynchronized();
+            _outputWriter.Write(_headerEncodingBuffer.AsSpan(0, payloadLength));
+        }
+        else
+        {
+            // This may happen in case of the TRAILERS after the initial encode operation.
+            while (writeResult == HPackHeaderWriter.HeaderWriteResult.BufferTooSmall)
             {
-                _outgoingFrame.ContinuationFlags = Http2ContinuationFrameFlags.END_HEADERS;
+                Debug.Assert(payloadLength == 0, "Payload written even though buffer is too small");
+                bufferSize *= 2;
+                largeHeaderBuffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+                buffer = largeHeaderBuffer.AsSpan(0, bufferSize);
+                writeResult = HPackHeaderWriter.RetryBeginEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, out payloadLength);
+                if (writeResult != HPackHeaderWriter.HeaderWriteResult.BufferTooSmall)
+                {
+                    SplitHeaderFramesToOutput(streamId, buffer[..payloadLength], endOfHeaders: writeResult == HPackHeaderWriter.HeaderWriteResult.Done, isFramePrepared: true);
+                }
+                ArrayPool<byte>.Shared.Return(largeHeaderBuffer);
             }
+            if (writeResult == HPackHeaderWriter.HeaderWriteResult.Done)
+            {
```

Comment:
Unfortunately, I don't think there's a "best" answer for how to manage oversized buffers without a detailed knowledge of how this will be used in practice.  On the one hand, we don't want to hold on to a really big buffer for the entire lifetime of the connection; on the other hand, a server that responds with one really big header will probably (we assume, but don't know) respond with other really big headers in the future.  In a way, I'd prefer to leave the decision to the pool (it's quite normal for pools to discard returned buffers they consider too large), but I appreciate that we have a lot more context here than the pool has.

I know that's not really what you were asking, but I wanted to provide some explanation for why my suggestions might not all trend in the same direction for this aspect of the change.

Reply:
Having said that, my intuition is that we want to reduce the number of times we replace the buffer in the course of serving a single response, so I think I'd rather not return the useful buffer immediately.  It actually looks like you may have done something more like that below?

In general, I agree that readability and maintainability is top priority, but I'm willing to make exceptions for perf (and security) and this is very performance-sensitive code (though, apparently, we didn't notice a large regression 😆).

Reply:
Within the loop, I have to return the buffer, because the loop always gets a larger one. The question is if this loops completes and has a buffer then this very last iteration of this loop could skip returning the buffer, so that the 'code below' could continue using it.

But I think it should be OK to return.

Reply:
I'd lean the other way (keep the buffer), but I don't feel strongly about it.

Reply:
I would like to keep it for now as-is. The size is preserved within the writer, so my main concern of having 2 loops increasing the buffer size over-and-over is less of a concern as it starts renting already a larger sized buffer.